feat: 2.0.0 - Merged multi-Zigbee2MQTT

Nerivec opened the pull request:
* feat: merge multiple API instances into a single interface
  * use tab design wherever merged does not make sense: Network, Logs, Settings
  * quick identification with a set of dots colored based on the API URL index (static). (Note: the dot also changes in case the underlaying WebSocket has a problem. )
  * Note: this remains mostly hidden in case of single-API use
* feat: [perf] batch log and device state updates using `requestAnimationFrame`
* feat: use notifications drawer instead of toasts for log-derived notifications
  * allows quick checking while remaining problems-free for screen space (especially with new multi-instance)
  * a dot will appear next to the notification icon if an `error`-level notification was received, as well as if an instance requires restarting (e.g. to apply settings)
* fix: use toasts only for success/error feedback in user-driven requests (`bridge/request`, `/get` & `/set`)
  * `/get` & `/set` errors are parsed to only show the actual error in the toast (full message available in notifications drawer & Logs page)
* fix: remove log lines in Dev Console in favor of new notifications drawer
* fix: "removed" navigation dead-end issues due to multi-instances switching with new merged design
* fix: move Frontend local storage settings into its own page
* fix: OTA table not using correct data when selecting all after filtering
* fix: input validation for scene, group
* fix: add log message on WebSocket error
* fix: improve handling of bad URLs (redirect as appropriate if missing segments or unknown device/group)
* feat: add [Docker `edge` version](https://github.com/Nerivec/zigbee2mqtt-windfront/pkgs/container/zigbee2mqtt-windfront) based on last `main` branch commit.

> [!IMPORTANT]
> BREAKING:
> - using multiple API instances in standalone will now be using the merged interface, no longer need switching
> - some URL path changes to accommodate new merged multi-instances support
>   - e.g. `#/device/0xbc33acfffe17628b` is now `#/device/0/0xbc33acfffe17628b` with the extra path used for API source indexing (as `[0-9]+`, always `0` in single-API use)
> - toasts no longer used for same purpose (see above)

Pinging some multi-instances users if I can get some in-depth testing on this and feedback on the design 😉 
@thk-socal @ams2990 @KoalaWerewolf @thargy 

### What to lookout for when testing this?

- any improvements you can fathom on the _newly introduced_ logic or design
- parts of the UI not updating properly when something _should change_
- performance issues (especially with larger multi-instances) => _make sure to identify if this affects a specific page, or several_

Nerivec commented:
As soon as it is merged yes. I'll add some details in the first post for testers to focus on (at least in the beginning).

thargy commented:
Edge AddOn Auto-Update doesn't work, indeed I couldn't get HA to detect an update, instead I had to uninstall and reinstall AddOn.

Love the combined interface! Literally a dream come true! Even if you change nothing, I'd still be happy, but here's my constructive feedback...

API names are much less discoverable. It would be great to add tooltips for desktop use (e.g., on the connection column, allow-join drop-down). Took me a while to figure out what the new notification menu was, and it's the only way to find which colour is for which connection. Can I suggest you add a multi-select dropdown in the connection column header that allows you to filter by collection and simultaneously see which colour is which API (similar to the search boxes in the other headers)? Similarly, perhaps add the API to the details 'About' tab.

In the 'Allow Join' dropdown having a bunch of 'All's at the top is less useful than perhaps 'All Local', 'All Upstairs', 'All Downstairs', 'All Extension', etc. which would again increase visibility of colour<->API correspondence.

The clear filter icon against the column searches isn't immediately obvious. Can I suggest you replace it with a classic 'x' and either disable/hide it when the filter is empty?

It would be useful to have filters for the other columns too. This can be done creatively. For example, multi-selects for connection, LQI (Red <= 74?, 74 < Yellow <= 126, White > 126), Last Seen (Just now, < 1 min, 1 - 10 min,  10min-1hr, 1hr-1day, > 1 day), Availability (Online, Offline, Disabled). If you can find a useful multiselect control it would also make the lining up more visually pleasing.


And it's never gonna happen, but I can't resist asking for my number one remaining feature request, and that's groups of groups (hierarchy) - but that requires Z2m support, and needs it to maintain zigbee groups automatically (i.e. expanding out a child group). With the new merged UI, it would be awesome if it could also handle cross API grouping. I currently have 64 groups across the APIs, like All Downstairs, All Upstairs, etc. which are then grouped in HA to 'All'. But each of those groups has to manually group, the child groups.

Nerivec commented:
@KoalaWerewolf 
Tables and the Dashboard page are now lazy-loaded to allow for better performance (multi-instance can easily reach 300+ devices now, so, we had to do something to prevent rendering a boatload of items at once that created far too long page load times).
This does change the scrolling behavior from before a bit, as items are added when the page's bottom is near-reached.

Can you elaborate on what you mean by "difficult", is it the behavior change that caught you off-guard, or do you see actual problems with the scrolling (not loading properly, etc.)?

Nerivec commented:
@ams2990 
> I'm not observing the icon change you mention, or the new option in settings.

That's in the frontend-specific settings (the display icon next to `Settings` in the navbar). There's a new "Multi-Instance" at the bottom.
It should show the names next to the colored dots by default, which can be changed there (note: some places have overrides to make more sense).

KoalaWerewolf commented:
> Can you elaborate on what you mean by "difficult", is it the behavior change that caught you off-guard, or do you see actual problems with the scrolling (not loading properly, etc.)?

The loading of devices seems to trigger a reload of the page and the scrolling starts from the beginning again. 

I recorded the behavior but I haven’t got the file below 10MB yet.

thargy commented:
## Works great

Show source name is great!
'x' for clear filter (with disable) is way clearer.

## Issues

I am on the latest release. I have 264 devices testing on desktop. The devices load, but it begins flickering between page loaded (scrollbar) to just 16 devices and back again repeatedly for a few seconds (~10s, but sometimes much shorter/longer) before settling on only showing the first 16 devices. On refresh, it sometimes only shows 16 devices, no scrolling.

If I try to scroll on each flicker (approx. every 300ms but random) it jumps back to the start.

On mobile (HA app), the flickering starts when I scroll to the end of the first page.

As such the Devices page is pretty unusable.

I haven't noticed a similar issue on groups, which seems to lazy load correctly (more testing needed)

I have noticed that only the first 16 items load on OTA (no flickering first).

In each case, changing the sort changed the 16 items shown as expected.

Logs, I am also seeing weird and inconsistent behaviour when switching between tabs. When I have a tab with relatively high churn, switching to another doesn't clear the logs from the previous tab.

## Suggestions

WIth the loss of the numbering column, and the inability to filter based on API, there is no longer an easy way to see the device count per coordinator.

Would still like to see source details in 'info' tab of device (when multi-source), there's space there to show color, API_NAME and API_URL. These should also be added to the Settings->About tab (at least the API_URL).

Auto API_NAME, as we're using the name in more places on the UI, creating a name automatically when one isn't supplied would be a good idea. The easiest way would be to take just the hostname from the URL.  Where duplicates are found, you can continue to add components until they are distinguished (e.g. port, and step along path). 

An alternative is to enforce naming, for example by changing the config to accept key-value pairs, e.g `Local="192.168.86.48:8123/api/hassio_ingress/i3LureG_jqGveD6qvlUs23KxUWnEa5j-TKsGwPCKEoU/api",Upstairs="192.168.86.42:8099/api"`

Enforce API URLs wrapped in double quotes `"` as they are not valid URL characters, unlike `,`, which is valid in a URL and so not a good separator. See [Section 2 of RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-2), which allows the following 84 characters:
`ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~:/?#[]@!$&'()*+,;=`

## Network map

I struggle to understand the map logic, in particular the circles. Sometimes the circles have a device name, other times it's the device name + 'Children', sometimes both exist for a device. The circles have one or more devices inside them, so they appear to be a grouping; however, that doesn't work when you switch to 3D mode (presumably they should be spheres in 3D, but they are circles on a single plane). It appears Routers are in blue, endpoint devices (I only have a couple!) are in green, and yellow is the coordinator, but a key would help.

Some parent edges appear to have arrows at both ends. What does that mean? 

When changing which edges are shown the node sizes can increase and overlap the arrow heads.

What does the arrow head indicate, I would expect one of the following (in preference order):
1) Parent edge, arrow head at parent (only)
2) Child edge, arrow head at parent (only)
3) Sibling edge, arrow head at both ends/no arrow head.

or
1) Parent edge, arrow head at child (only)
2) Child edge, arrow head at parent (only)
3) Sibling edge, arrow head at both ends/no arrow head.

or
1) Parent edge, arrow head at parent (only)
2) Child edge, arrow head at child (only)
3) Sibling edge, arrow head at both ends/no arrow head.

I can't tell which is true, and so a legend would help.

Nerivec commented:
Fantastic 😃
I'll go over these after I'm done with the perf issues (I've changed the method for the dashboard & co, now need to do the tables, as previous logic appeared to cause problems with some devices/browsers).

PS: there's a small legend already, above the network map. For the arrows, e.g. "[child] EndDevice -> Coordinator", as in "EndDevice is child of Coordinator", e.g. "[parent] Coordinator -> Router", as in "Coordinator is parent of Router". The big problem is, this data relies of what devices report, and often enough, that data can be stale, or even plain wrong (Tuya...), so connections sometimes don't make sense... _There isn't much we can do about that though, as "correcting" these would raise the issue of "which is right"._

KoalaWerewolf commented:
Here's a recording of scrolling in Firefox on Mac OS. I've removed Friendly Name for privacy, you can still easily see that the page resets to the top continually. I updated the Add-on earlier today, AFAIK this is the latest code.

https://github.com/user-attachments/assets/7bddb42c-4123-48bf-86ad-72db338878db

thargy commented:
> If I can get some quick feedback on the scrolling issues with latest `edge` before I move on to other points. I changed the technique, used a well-established library, hopefully should fix both scrolling and performance issues. Bonus, the Dashboard gets a real masonry layout, which better fits with the varying content caused by devices with small/large feature amounts.

Nailed it. Tested on edge, and HA iOS app. Scrolling is excellent on both. There is a barely perceptible delay as it loads data on edge whilst scrolling quickly, but it is hardly noticeable, on iOS it is invisible. Performance is stunning. Love the new masonry layout on the dashboard too.

This is going to be a real game-changer when it goes to prod!

KoalaWerewolf commented:
Works great!

Nerivec commented:
@thargy 
3. this is trickier here since it uses a static site. I'll see if there's a way to do this uniformly across platforms. Note: in Z2M it shows "unknown" for release, but should show the proper commit for dev/edge (except for specific cases where git isn't available)
4. should be _mostly_ fixed with the new filtering abilities
5. fixed
6. similar problem as 3, quite limited in what's doable & presentable across platforms.
7. in force layouts, the devices are clustered by their parents, which I believe is what you meant by circles. In 3d, the cluster renders as a sort of "tunnel". It determines the parent for clustering from either parent or child relationships. In the latter, it indicates "children".
8. this is in the legend as mentioned above (PS; you can also swap to icons now)
10. for detailed views (tables), I think it's best to keep the IEEE visible by default, users can remove it if they want
11. 
  - some browsers show their own "clear" version independently of the design (with input type "search"). I can't remove the "actual clear" since some browsers don't show one. And using "text" type instead of "search" would lose the ability to clear it with ESC (that is supported in most browsers afaik), and whatever other mechanisms users of specific devices might be used to.
  - just added more filters to provide alternatives to what is doable in the Devices page

Nerivec commented:
@0rsa I saw you were using multi-instance. A lot has happened in that area since last release, if you want to give current `edge` a try (not sure which method you are using, but both Docker & HA add-on have an `edge` version) and provide some quick feedback here with the rest. 👍

ams2990 commented:
> Switching between Logs tabs changes which instance's logs appear at the bottom as the newest logs, but does not clear already-loaded logs, producing a mishmash of logs from different sources.

I just updated to the latest edge, and I'm no longer able to reproduce this. When switching log sources, all of the log lines are replaced as expected, rather than getting some from one source and some from another.
